enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,344 @@
+
+### Non-Goals
+
+- Provide each pod in a StatefulSet with a separate secret containing only the certificate+key
+for that single pod:
+
+  This is not currently possible,
+  because the StatefulSet pod template does not have any templating mechanism
+  that would allow a ${podID}-dependent value to be used for a secret name or a volume mount;
+  there’s only StatefulSet.spec.volumeClaimTemplates , which is not practical to use.
+
+  Upstream I could find https://github.com/kubernetes/kubernetes/issues/40651 , which only ends up
+  discussing environment variables (which work via the downward API).
+
+  The per-${podID} behavior can be simulated using an init container
+  (e.g. https://itnext.io/kubernetes-statefulset-initialization-with-unique-configs-per-pod-7e02c01ada65 ),
+  by making a full set of certificates available to all pods
+  and relying on an init container to choose one certificate and make that one
+  (and none of the others) available to the actual server container.
+  
+  That’s cumbersome, but ultimately probably secure enough — if it makes sense to protect
+  the individual StatefulSet pods from each other at all.
+
+- Provide certificates for `127.0.0.1`, `localhost`, or NodeIP service endpoints:
+
+  Giving out certificates from widely-accepted CAs (like the service CA)
+  with these names makes no sense because they don’t specify any service identity,
+  e.g. any compromised serviceA would get a “localhost” certificate
+  that is acceptable for connections from a victim serviceB.
+
+  (Reportedly some OSes do actually query `localhost` on DNS on the net:
+  https://archive.cabforum.org/pipermail/public/2015-June/005673.html )
+
+  If a client wants to connect to _only_ a “true localhost” service,
+  a service CA-signed certificate does not provide that property.
+  The best way to do that is to do it directly inside the process,
+  without using TLS (and the associated cryptography overhead) at all.
+  Second best is to rely on known properties of the environment,
+  i.e. to assume that 127.0.0.1 is localhost-only
+  and to use a non-TLS connection on that address.
+
+  If a _pod_ had to use TLS for localhost for some reason, the closest to a right way to do that
+  (relying on TLS to verify that only loopback connections happen) is to:
+  - Create a temporary CA (specific to the individual pod for)
+  - Generate, and have that CA sign a localhost certificate (specific to the individual pod)
+  - Set up the CA certificate as a trusted root
+  - _Irreversibly erase_ the CA private key
+  - Set up the pod with the generated localhost certificate+private key.
+  - (This doesn’t benefit from service-ca involvement at all.)
+  
+  Similarly single-purpose CA could be set up for NodeIP services.
+
+- Generate a single certificate that signs multiple ~unrelated services as subjects
+  (combining the StatefulSet per-pod subjects and ClusterIP services,
+  or a ClusterIP service with NodePort service):
```

Comment:
one could create a separate cert for StatefulSet backed ClusterIP service, right? I.e. one that does has per-pod hostname, but prely the service IP and service DNS name as subjects.

Reply:
I don’t understand.

[For context, a minimal certificate is basically a tuple (subject public key, subject identity, issuer’s signature of the two)]. Within a certificate, the hostname and the subject are ~the same thing (whether in the subject DN, or in the subjectAltName extension), so I don’t know what having “purely the service DNS name as subject” but a “per-pod host name” would look like.

Yes, one can create a set of separate certificates:
- $statefulSet.Spec.Replicas separate certificates, each with a single per-pod host name
- A single certificate with the host name of ClusterIP $serviceA
- A single certificate with the host name of ClusterIP $serviceB

and a single pod can accept connections for all three of these (by configuring $serviceA and $serviceB selectors to use that pod, opening / mapping ports, etc.); that pod then has to be implemented so that it serves the appropriate certificate for each connection. That’s what this design recommends.

The user request, and something this proposal does not do, is to provide a single certificate that has all of these subjects, so that the pod can just serve a single certificate on all open ports and not have to differentiate.